File: sistemas-controle-2/p1-content/Projeto_PID.py
# -*- coding: utf-8 -*-
"""
Created on Thu Sep 10 17:48:34 2020

@author: rcard
"""

#As próximas 3 linhas são para selecionar entre plot inline ou em nova janela
#Útil para rlocus
from IPython import get_ipython
get_ipython().run_line_magic('matplotlib', 'qt')
#get_ipython().run_line_magic('matplotlib', 'inline')

import numpy as np                #Biblioteca para cálculo numérico
import math                       #Funções matemáticas
import matplotlib.pyplot as plt   # Funções de plot similares as do MATLAB
import control as ctrl            # Biblioteca para controle
from control.matlab import *      # Funções para controle similares as do MATLAB

#Planta
num=[4]
den=[1, 0.5, 0]
G=tf(num,den) 
print(G)

#Planta sem compensação em malha fechada
Gmf=feedback(G,1)
#Polos de MF sem compensação
print(pole(Gmf))

#Compensador
numc=[1.13, 6.4749, 1.2498]
denc=[1, 0]
Gc=tf(numc,denc)
print(Gc)
#Planta com compensação em malha fechada
Gmfc=feedback(Gc*G,1)
#Polos de MF do sistema compensado
print(pole(Gmfc))

#Resposta ao degrau
t=np.linspace(0,25,1000)
y1, t1 = step(Gmf,t)
y2, t2 = step(Gmfc,t)
plt.figure()
plt.plot(t1,y1,t2,y2)
plt.title('Resposta ao degrau')
plt.legend(('Sem compensação', 'Com compensação'))
plt.xlabel('t(s)')
plt.ylabel('Amplitude')
plt.grid()

#Lugar das raízes da planta e da planta compensada
rlocus(G)
rlocus(Gc*G)


#Resposta a rampa
Gi=tf([1],[1, 0]);
T=np.linspace(0,50,1000)
y3, t3 = step(Gi,T)
y4, t4 = step(Gi*Gmf,T)
y5, t5 = step(Gi*Gmfc,T)
plt.figure()
plt.plot(t3,y3,t4,y4,t5,y5)
plt.title('Resposta a rampa')
plt.legend(('Referência', 'Gmf','Gmfc'))
plt.xlabel('t(s)')
plt.ylabel('Amplitude')
plt.grid()

#Erros
y6, t6 = step(Gi,T)
y7, t7 = step(Gi*Gmf,T)
y8, t8 = step(Gi*Gmfc,T)
erro_MF=y3-y4
erro_MF_C=y3-y5

plt.figure()
plt.plot(T,erro_MF,T,erro_MF_C)
plt.title('Erro de rastreamento para referência do tipo rampa')
plt.legend(('Sem compensação', 'Com compensação'))
plt.xlabel('t(s)')
plt.ylabel('Amplitude')
plt.grid()

File: sistemas-controle-2/p1-content/Projeto_PID.m
close all;
clear all;
clc;

%Planta
num=4;
den=[1 0.5 0];
G=tf(num,den);
%Planta sem compensação em malha fechada
Gmf=feedback(G,1);

%Compensador
numc=1.13*[1 5.73 1.106];
denc=[1 0];
Gc=tf(numc,denc);
%Planta com compensação em malha fechada
Gmfc=feedback(Gc*G,1);
%Polos de MF do sistema compensado
pole(Gmfc)

%Resposta ao degrau
figure;
step(Gmf,30);
hold;
step(Gmfc,30);
legend('Sem compensação', 'Com compensação');

%Lugar das raízes da planta e planta compensada
figure;
rlocus(G);
legend('G(s)');
figure;
rlocus(Gc*G);
legend('Gc(s)G(s)');

%Resposta a rampa
Gi=tf(1,[1 0]);
figure;
step(Gi,'b',15);
hold;
step(Gi*Gmf,'r',15);
step(Gi*Gmfc,'k',15);
title('Ramp Response');
legend('Referência','Gmf','Gmfc');

%Erros
[Yi t]=step(Gi,'b',linspace(0,50,1000));
[Y1 t]=step(Gi*Gmf,'r',linspace(0,50,1000));
[Y2 t]=step(Gi*Gmfc,'k',linspace(0,50,1000));

erro_MF=Yi-Y1;
erro_MF_C=Yi-Y2;

%Exibe erros
figure;
plot(t,erro_MF,'r');
hold;
plot(t,erro_MF_C,'b');
legend('Erro sem compensador','Erro com compensador');
xlabel('t (s)');
ylabel('Amplitude');